README.rst
repocutter
==========
.. image:: https://img.shields.io/badge/License-MIT-yellow.svg
    :target: https://opensource.org/licenses/MIT
    :alt: License
.. image:: https://img.shields.io/pypi/v/repocutter
    :target: https://pypi.org/project/repocutter/
    :alt: PyPI
.. image:: https://github.com/jshwi/repocutter/actions/workflows/build.yaml/badge.svg
    :target: https://github.com/jshwi/repocutter/actions/workflows/build.yaml
    :alt: Build
.. image:: https://github.com/jshwi/repocutter/actions/workflows/codeql-analysis.yml/badge.svg
    :target: https://github.com/jshwi/repocutter/actions/workflows/codeql-analysis.yml
    :alt: CodeQL
.. image:: https://results.pre-commit.ci/badge/github/jshwi/repocutter/master.svg
   :target: https://results.pre-commit.ci/latest/github/jshwi/repocutter/master
   :alt: pre-commit.ci status
.. image:: https://codecov.io/gh/jshwi/repocutter/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/jshwi/repocutter
    :alt: codecov.io
.. image:: https://readthedocs.org/projects/repocutter/badge/?version=latest
    :target: https://repocutter.readthedocs.io/en/latest/?badge=latest
    :alt: readthedocs.org
.. image:: https://img.shields.io/badge/python-3.8-blue.svg
    :target: https://www.python.org/downloads/release/python-380
    :alt: python3.8
.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black
    :alt: Black
.. image:: https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336
    :target: https://pycqa.github.io/isort/
    :alt: isort
.. image:: https://img.shields.io/badge/%20formatter-docformatter-fedcba.svg
    :target: https://github.com/PyCQA/docformatter
    :alt: docformatter
.. image:: https://img.shields.io/badge/linting-pylint-yellowgreen
    :target: https://github.com/PyCQA/pylint
    :alt: pylint
.. image:: https://img.shields.io/badge/security-bandit-yellow.svg
    :target: https://github.com/PyCQA/bandit
    :alt: Security Status
.. image:: https://snyk.io/test/github/jshwi/repocutter/badge.svg
    :target: https://snyk.io/test/github/jshwi/repocutter/badge.svg
    :alt: Known Vulnerabilities
.. image:: https://snyk.io/advisor/python/repocutter/badge.svg
  :target: https://snyk.io/advisor/python/repocutter
  :alt: repocutter

Checkout repos to current cookiecutter config
---------------------------------------------

Checkout one or more repos to current `cookiecutter <https://github.com/cookiecutter/cookiecutter>`_ config

This will make changes to local repositories, hopefully preserving their history

Ideally only the working tree will change

Ignored files should be backed up

Use with caution

Usage
-----

.. code-block:: console

    usage: repocutter [-h] [-v] [-a] [-c] [-b REV,NEW] [-i LIST] PATH [REPOS [REPOS ...]]

    Checkout repos to current cookiecutter config

    positional arguments:
      PATH                          path to cookiecutter template dir
      REPOS                         repos to run cookiecutter over

    optional arguments:
      -h, --help                    show this help message and exit
      -v, --version                 show program's version number and exit
      -a, --accept-hooks            accept pre/post hooks
      -c, --gc                      clean up backups from previous runs
      -b REV,NEW, --branch REV,NEW  checkout new branch from existing revision
      -i LIST, --ignore LIST        comma separated list of paths to ignore, cookiecutter vars are allowed

Configuration
-------------

Currently only written for a configuration exactly like below

Technically a repo would not need to be a `poetry <https://github.com/python-poetry/poetry>`_ project if the below section exists within its pyproject.toml file

This is the only use case at this time (If there are any other configurations you would like added please leave an `issue <https://github.com/jshwi/repocutter/issues>`_)

Each repository's pyproject.toml file will be parsed for data to recreate its working tree

A ``poetry`` section in the project's pyproject.toml file that looks like the following...

.. code-block:: toml

    [tool.poetry]
    description = "Checkout repos to current cookiecutter config"
    keywords = [
      "config",
      "cookiecutter",
      "jinja2",
      "repo",
      "template"
    ]
    name = "repocutter"
    version = "0.2.0"

...will temporarily write to the ``cookiecutter`` project's cookiecutter.json file until the repo is created

.. code-block:: json

    {
      "project_name": "repocutter",
      "project_version": "0.2.0",
      "project_description": "Checkout repos to current cookiecutter config",
      "project_keywords": "config,cookiecutter,jinja2,repo,template",
    }

The above configuration will reduce the diff, but it will still work if your config is not exactly the same

Why?
----
As time goes on, and you use ``cookiecutter`` for new projects, you will make more and more changes to your ``cookiecutter`` repo

You will find these new project layouts are preferable to your older, more outdated, projects

If you have a project layout configured with ``cookiecutter`` then it's likely you will want this layout for all your projects

Configuring your existing projects manually is even more tedious than configuring a new project manually, especially if you have a lot of them

By checking out your projects to your configured ``cookiecutter`` layout, you can use whatever diff tool you use to rollback any undesired changes


CHANGELOG.md
Changelog
=========
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

[Unreleased](https://github.com/jshwi/repocutter/compare/v0.6.2...HEAD)
------------------------------------------------------------------------

[0.6.2](https://github.com/jshwi/repocutter/releases/tag/v0.6.2) - 2023-09-30
------------------------------------------------------------------------
### Security
- update dependencies

[0.6.1](https://github.com/jshwi/repocutter/releases/tag/v0.6.1) - 2023-01-24
------------------------------------------------------------------------
### Fixed
- Disable git hooks

[0.6.0](https://github.com/jshwi/repocutter/releases/tag/v0.6.0) - 2023-01-24
------------------------------------------------------------------------
### Fixed
- Evaluate template variables in cookiecutter.json
- Fix error raised if already on revision with `-b/--branch`

[0.5.0](https://github.com/jshwi/repocutter/releases/tag/v0.5.0) - 2023-01-22
------------------------------------------------------------------------
### Added
- Add `-b/--branch` argument

[0.4.0](https://github.com/jshwi/repocutter/releases/tag/v0.4.0) - 2023-01-21
------------------------------------------------------------------------
### Added
- Add `-i/--ignore` argument

[0.3.0](https://github.com/jshwi/repocutter/releases/tag/v0.3.0) - 2023-01-15
------------------------------------------------------------------------
### Added
- Add `-c/--gc` argument

[0.2.0](https://github.com/jshwi/repocutter/releases/tag/v0.2.0) - 2023-01-15
------------------------------------------------------------------------
### Added
- Add report
- Add temporary staging environment

### Changed
- Continue if repo provided does not have a pyproject.toml file
- Continue if path provided is not a repository
- Continue if non-existing repo path provided

[0.1.0](https://github.com/jshwi/repocutter/releases/tag/v0.1.0) - 2023-01-14
------------------------------------------------------------------------
### Added
- Initial release


tests/_utils.py
"""
tests._utils
============
"""
# pylint: disable=too-few-public-methods
from __future__ import annotations

import contextlib
import json
import os
import typing as t
from pathlib import Path

from gitspy import Git as _Git
from templatest.utils import VarPrefix, VarSeq

FixtureMain = t.Callable[..., int]
FixtureMakeTree = t.Callable[[Path, t.Dict[t.Any, t.Any]], None]
FixtureGitInit = t.Callable[[Path], None]
FixtureMockCookiecutter = t.Callable[[t.Callable[..., None]], None]
FixtureMockTemporaryDirectory = t.Callable[..., None]
FixtureMakeRepos = t.Callable[..., t.List[Path]]

NAME = "name"

name = VarSeq(NAME)
name_dash = VarSeq(NAME, suffix="-")
description = VarSeq("description")
flags = VarPrefix("--", slug="-")
file = VarSeq("file")
folder = VarSeq("dir")
version = VarSeq("1.0", ".")
keyword = VarSeq("keyword")
tmp = VarSeq("tmp")

cookiecutter_json = {
    "author": "Stephen Whitlock",
    "github_username": "jshwi",
    "email": "stephen@jshwisolutions.com",
    "project_name": "project-name",
    "project_slug": "{{ cookiecutter.project_name|lower|replace('-', '_') }}",
    "project_version": "0.0.0",
    "project_description": (
        "Short description for {{ cookiecutter.project_name }}"
    ),
    "project_keywords": "comma,separated,list",
    "include_entry_point": ["n", "y"],
}


GIT_DIR = ".git"
GIT_TREE = {
    "config": None,
    "objects": {"pack": None, "info": None},
    "HEAD": None,
    "FETCH_LOG": None,
    "refs": {"heads": None, "tags": None},
    "FETCH_HEAD": None,
    "hooks": {
        "commit-msg": None,
        "post-checkout": None,
        "post-commit": None,
        "post-merge": None,
        "post-rewrite": None,
        "pre-commit": None,
        "pre-merge-commit": None,
        "prepare-commit-msg": None,
    },
}
PYPROJECT_TOML = "pyproject.toml"


class MockJson:
    """Mock ``json`` module."""

    def __init__(self) -> None:
        self._dumped: dict[str, str] = {}

    @property
    def dumped(self) -> dict[str, str]:
        """Return object that has been dumped."""
        return self._dumped

    def dumps(self, obj: dict[str, str]) -> str:
        """Mock ``json.dumps`` to capture what was dumped.

        :param obj: Object to dump.
        :return: str representation of dict as json.
        """
        self._dumped.update(obj)
        return json.dumps(obj)

    @staticmethod
    def loads(string: str) -> dict[str, str]:
        """Mock ``json.loads``.

        :param string: Json str.
        :return: Json str as dict object.
        """
        return json.loads(string)


class Git(_Git):
    """Mock ``Git`` object."""

    def __init__(self) -> None:
        super().__init__()
        self._called: list[str] = []

    def call(self, *args: str, **_: bool | str | os.PathLike) -> int:
        self._called.append(" ".join(str(i) for i in args))
        return 0

    @property
    def called(self) -> list[str]:
        """Get git commands that were called."""
        return self._called


class MockTemporaryDirectory:
    """Mock ``tempfile.TemporaryDirectory``.

    :param temp_dirs: Paths to mock ``tempfile.TemporaryDirectory``
        with.
    """

    def __init__(self, *temp_dirs: Path) -> None:
        self._temp_dirs = list(temp_dirs)

    @contextlib.contextmanager
    def open(self) -> t.Generator[str, None, None]:
        """Mock dir returned from ``TemporaryDirectory`` constructor.

        :return: Generator yielding self.
        """
        temp_dir = self._temp_dirs.pop(0)
        temp_dir.mkdir()
        yield str(temp_dir)


class KeyWords(t.List[str]):
    """Represents a list of keywords.

    :param suffix: Something to make the keywords unique.
    """

    def __init__(self, suffix: int) -> None:
        super().__init__([f"{keyword[suffix]}_{i}" for i in range(5)])


class PyProjectParams(
    t.Dict[str, t.Dict[str, t.Dict[str, t.Union[str, t.List[str], KeyWords]]]]
):
    """Represents contents of a pyproject.toml file.

    :param project_name: Name of project.
    :param project_description: Project description.
    :param project_keywords: Project keywords.
    :param project_version: Project's version.
    """

    authors = ["jshwi <stephen@jshwisolutions.com>"]

    def __init__(
        self,
        project_name: str,
        project_description: str,
        project_keywords: KeyWords,
        project_version: str,
    ) -> None:
        super().__init__(
            {
                "tool": {
                    "poetry": {
                        "authors": self.authors,
                        "description": project_description,
                        "documentation": (
                            f"https://{project_name}.readthedocs.io/en/latest"
                        ),
                        "homepage": (
                            f"https://pypi.org/project/{project_name}/"
                        ),
                        "keywords": project_keywords,
                        "license": "MIT",
                        "maintainers": self.authors,
                        NAME: project_name,
                        "readme": "README.rst",
                        "repository": "https://github.com/jshwi/repo1",
                        "version": project_version,
                    }
                }
            }
        )


class Repo(t.NamedTuple):
    """New repository type."""

    name: str
    contents: t.Dict[str, t.Any]


tests/__init__.py
"""
tests
=====

Test package for ``repocutter``.
"""


whitelist.py
FixtureGitInit  # unused variable (tests/_utils.py:19)
_.call  # unused method (tests/_test.py:649)
_.call  # unused method (tests/_utils.py:110)
exc_tb  # unused variable (repocutter/_main.py:162)
exc_type  # unused variable (repocutter/_main.py:160)
exc_val  # unused variable (repocutter/_main.py:161)
fixture_cache_dir  # unused function (tests/conftest.py:92)
fixture_cookiecutter_package  # unused function (tests/conftest.py:70)
fixture_environment  # unused function (tests/conftest.py:105)
fixture_main  # unused function (tests/conftest.py:118)
fixture_make_repo  # unused function (tests/conftest.py:48)
fixture_make_tree  # unused function (tests/conftest.py:29)
fixture_mock_cookiecutter  # unused function (tests/conftest.py:148)
fixture_mock_json  # unused function (tests/conftest.py:136)
fixture_mock_temporary_dir  # unused function (tests/conftest.py:164)


repocutter/_main.py
"""
repocutter._main
================

Contains package entry point.
"""
from __future__ import annotations

import contextlib as _contextlib
import json as _json
import os as _os
import re as _re
import shutil as _shutil
import typing as _t
from argparse import HelpFormatter as _HelpFormatter
from pathlib import Path as _Path
from subprocess import CalledProcessError as _CalledProcessError
from tempfile import TemporaryDirectory as _TemporaryDirectory
from types import TracebackType as _TracebackType

import appdirs as _appdirs
import checksumdir as _checksumdir
import tomli as _tomli
from arcon import ArgumentParser as _ArgumentParser
from cookiecutter.environment import StrictEnvironment as _StrictEnvironment
from cookiecutter.main import cookiecutter as _cookiecutter
from cookiecutter.prompt import render_variable as _render_variable
from gitspy import Git as _Git
from object_colors import Color as _Color

from ._version import __version__

_NAME = __name__.split(".", maxsplit=1)[0]
_GIT_DIR = ".git"
_INFO = 20
_WARNING = 30
_ERROR = 40
_PRE_COMMIT_CONFIG = ".pre-commit-config.yaml"
_HOOKS = "hooks"

_git = _Git()
_color = _Color()

_color.populate_colors()


class _Parser(_ArgumentParser):
    def __init__(self) -> None:
        super().__init__(
            __version__,
            prog=_color.cyan.get(_NAME),
            description="Checkout repos to current cookiecutter config",
            formatter_class=lambda prog: _HelpFormatter(
                prog, max_help_position=45
            ),
        )
        self._add_arguments()
        self.args = self.parse_args()
        self.args.repos = tuple(_Path(i) for i in self.args.repos)
        self.args.ignore = tuple(_Path(i) for i in self.args.ignore)

    def _add_arguments(self) -> None:
        self.add_argument(
            "path",
            action="store",
            type=_Path,
            metavar="PATH",
            help="path to cookiecutter template dir",
        )
        self.add_argument(
            "repos",
            nargs="*",
            action="store",
            metavar="REPOS",
            help="repos to run cookiecutter over",
        )
        self.add_argument(
            "-a",
            "--accept-hooks",
            action="store_true",
            help="accept pre/post hooks",
        )
        self.add_argument(
            "-c",
            "--gc",
            action="store_true",
            help="clean up backups from previous runs",
        )
        self.add_list_argument(
            "-b",
            "--branch",
            action="store",
            metavar="REV,NEW",
            help="checkout new branch from existing revision",
        )
        self.add_list_argument(
            "-i",
            "--ignore",
            action="store",
            metavar="LIST",
            help=(
                "comma separated list of paths to ignore, cookiecutter vars"
                " are allowed"
            ),
        )


class _MetaData(_t.Dict[str, str]):
    _keys = "name", "version", "description", "keywords"

    @staticmethod
    def _format(key: object) -> str:
        if isinstance(key, list):
            return ",".join(key)

        return str(key)

    def __init__(self, __m: dict[str, dict[str, dict[str, str]]]) -> None:
        super().__init__(
            {
                f"project_{k}": self._format(__m["tool"]["poetry"][k])
                for k in self._keys
            }
        )

    def setentry(self, repo: _Path) -> None:
        """Set the ``include_entry_point`` key.

        :param repo: Path to repository.
        """
        main_file = repo / self["project_name"] / "__main__.py"
        self["include_entry_point"] = "y" if main_file.is_file() else "n"


class _Defaults(_t.Dict[str, _t.Union[str, _t.List[str]]]):
    def __init__(self, __m: dict[str, str | list[str]]) -> None:
        super().__init__(__m)
        self._env = _StrictEnvironment(context=self)

    def render(self, metadata: _MetaData) -> None:
        """Update and render `jinja2` template vars with parsed data.

        :param metadata: Data parsed from config.
        """
        self.update(metadata)
        for key, value in self.items():
            self[key] = _render_variable(self._env, value, self)


class _ChDir:
    def __init__(self, path: _Path) -> None:
        self._cwd = _Path.cwd()
        _os.chdir(path)

    def __enter__(self) -> _ChDir:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: _TracebackType | None,
    ):
        _os.chdir(self._cwd)


@_contextlib.contextmanager
def _temporary_directory() -> _t.Generator[_Path, None, None]:
    with _TemporaryDirectory() as tempdir:
        yield _Path(tempdir)


def _report(level: int, repo: _Path, message: str) -> None:
    ident = (15 - len(str(repo))) * " "
    color = {_INFO: _color.green, _WARNING: _color.yellow, _ERROR: _color.red}
    print(f"[{color[level].get(repo)}{ident}] {message}")


def _garbage_collection(cache_dir: _Path) -> None:
    if cache_dir.is_dir():
        _report(_INFO, cache_dir, "cleaning")
        for path in cache_dir.iterdir():
            _shutil.rmtree(path)


@_contextlib.contextmanager
def _stage_pre_commit(repo: _Path) -> _t.Generator[None, None, None]:
    path = repo / _PRE_COMMIT_CONFIG
    if path.is_file():
        _git.add(path.name)

    yield
    _git.reset(path.name, file=_os.devnull)


@_contextlib.contextmanager
def _disable_hooks(repo: _Path) -> _t.Generator[None, None, None]:
    path = repo / _GIT_DIR / _HOOKS
    with _temporary_directory() as temp:
        temp_hooks = temp / _HOOKS
        if path.is_dir():
            _shutil.move(str(path), temp_hooks)

        yield
        if temp_hooks.is_dir():
            _shutil.move(str(temp_hooks), path)


def _revert_ignored(
    paths: tuple[_Path, ...], repo: _Path, defaults: _Defaults
) -> None:
    with _ChDir(repo):
        with _stage_pre_commit(repo):
            for path in paths:
                rgx = _re.match(r"{{\s?cookiecutter\.(.[^ ]*)\s?}}", str(path))
                if rgx:
                    new_path = defaults.get(rgx.group(1))
                    if not isinstance(new_path, list) and new_path is not None:
                        path = _Path(new_path)

                if path.exists():
                    if path.is_dir():
                        _shutil.rmtree(path)
                    else:
                        _os.remove(path)

                try:
                    _git.checkout("HEAD", "--", str(path), file=_os.devnull)
                except _CalledProcessError:
                    pass


def _checkout_branches(repo: _Path, existing: str, new: str) -> int:
    try:
        _git.checkout(existing, capture=True)
        _git.stderr()  # flush stderr if non-fatal
        _git.checkout("-b", new, capture=True)
    except _CalledProcessError:
        _report(_WARNING, repo, _git.stderr()[0].split(": ")[1])
        return 1

    return 0


def main() -> int:
    """Main function for package.

    :return: Exit status.
    """
    parser = _Parser()
    cache_dir = _Path(_appdirs.user_cache_dir(_NAME))
    if parser.args.gc:
        _garbage_collection(cache_dir)

    readable_config = parser.args.path / "cookiecutter.json"
    with _temporary_directory() as temp:
        template = temp / parser.args.path.name
        _shutil.copytree(parser.args.path, template)
        writeable_config = template / "cookiecutter.json"
        for repo in parser.args.repos:
            defaults = _Defaults(
                _json.loads(readable_config.read_text(encoding="utf-8"))
            )
            pyproject_toml = repo / "pyproject.toml"
            git_dir = repo / _GIT_DIR
            if not repo.is_dir():
                _report(_WARNING, repo, "does not exist")
                continue

            if not git_dir.is_dir():
                _report(_WARNING, repo, "not a repository")
                continue

            if not pyproject_toml.is_file():
                _report(_WARNING, repo, "missing pyproject.toml")
                continue

            temp_repo = temp / repo.name
            _shutil.copytree(repo, temp_repo)
            with _disable_hooks(temp_repo):
                archive_name = f"{repo.name}-{_checksumdir.dirhash(git_dir)}"
                archived_repo = cache_dir / archive_name
                metadata = _MetaData(
                    _tomli.loads(pyproject_toml.read_text(encoding="utf-8"))
                )
                metadata.setentry(repo)
                defaults.render(metadata)
                writeable_config.write_text(
                    _json.dumps(defaults), encoding="utf-8"
                )
                temp_git_dir = temp_repo / _GIT_DIR
                with _ChDir(temp_repo):
                    _git.stash(file=_os.devnull)
                    if parser.args.branch and _checkout_branches(
                        repo, parser.args.branch[0], parser.args.branch[1]
                    ):
                        continue

                if archived_repo.is_dir():
                    _shutil.rmtree(archived_repo)

                _shutil.move(str(temp_repo), archived_repo)
                _cookiecutter(
                    template=str(template),
                    output_dir=str(temp),
                    no_input=True,
                    accept_hooks=parser.args.accept_hooks,
                )
                if temp_git_dir.is_dir():
                    _shutil.rmtree(temp_git_dir)

                _shutil.copytree(archived_repo / _GIT_DIR, temp_git_dir)
                _revert_ignored(parser.args.ignore, temp_repo, defaults)
                _shutil.rmtree(repo)
                _shutil.move(temp_repo, repo)

                _report(_INFO, repo, "success")

    return 0
